File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
# 랜선 개수, 필요한 랜선 개수
K, N = map(int, input().split(" "))
l = []

for _ in range(K):
    l.append(int(input()))

l.sort()

start = 1
end = l[-1]
cable_num=0
while start<=end:
    '''
    등분할 길이가 mid가 된다.
    등분할 길이가 크면 케이블 개수가 적어진다.
    그래서 케이블 개수가 많으면 등분할 길이를 크게해야 한다.
    '''
    mid = (start + end) // 2
    cable_num = 0
    for i in range(len(l)):
        cable_num += l[i] // mid
    if cable_num >= N:
        start = mid + 1
        ans = mid
    else:
        end = mid - 1

print(ans)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
#include<iostream>
#include<vector>

using namespace std;
/* 랜선 자르기

1. K개의 랜선으로 N 개의 랜선을 만들어야 한다.
2. 이때 만들 수 있는 랜선 길이의 최대 길이를 구하라
*/



int main() {

	cin.tie(0);
	cout.tie(0);
	int n, k, tmp;
	long long end = 0;
	vector<int>v;
	cin >> n>>k;
	for (size_t i = 0; i < n; ++i) {
		cin >> tmp;
		v.push_back(tmp);
		if (tmp > end) {
			end = tmp;
		}
	}

	long long start = 1;
	long long mid=0;
	int ans=1;
	while (start <= end) {
		mid = (start + end) / 2;
		long long lanNum=0;
		for (size_t i = 0; i < n; ++i) {
			lanNum+=v[i] / mid;
		}
		if (lanNum >= k) {
			ans = mid;
			start = mid + 1;
		}
		else {
			end = mid - 1;
		}
	}

	cout << ans;
	return 0;
}